lib/vmfloaty/abs.rb

```diff
@@ -0,0 +1,257 @@
+# frozen_string_literal: true
+
+require 'vmfloaty/errors'
+require 'vmfloaty/http'
+require 'faraday'
+require 'json'
+
+class ABS
+  # List active VMs in ABS
+  # This is what a job request looks like:
+  # {
+  #   "state":"filled",
+  #   "last_processed":"2019-10-31 20:59:33 +0000",
+  #   "allocated_resources": [
+  #     {
+  #       "hostname":"h3oyntawjm7xdch.delivery.puppetlabs.net",
+  #       "type":"centos-7.2-tmpfs-x86_64",
+  #       "engine":"vmpooler"}
+  #     ],
+  #   "audit_log":{
+  #     "2019-10-30 20:33:12 +0000":"Allocated h3oyntawjm7xdch.delivery.puppetlabs.net for job 1572467589"
+  #   },
+  #   "request":{
+  #     "resources":{
+  #       "centos-7.2-tmpfs-x86_64":1
+  #     },
+  #   "job": {
+  #     "id":1572467589,
+  #     "tags": {
+  #           "user":"mikker",
+  #           "url_string":"floaty://mikker/1572467589"
+  # },
+  #           "user":"mikker",
+  #           "time-received":1572467589
+  #         }
+  #       }
+  #     }
+  #
+
+  @active_hostnames = {}
+
+  def self.list_active(verbose, url, _token, user)
+    all_jobs = []
+    @active_hostnames = {}
+
+    get_active_requests(verbose, url, user).each do |req_hash|
+      all_jobs.push(req_hash['request']['job']['id'])
+      @active_hostnames[req_hash['request']['job']['id']] = req_hash
+    end
+
+    all_jobs
+  end
+
+  def self.get_active_requests(verbose, url, user)
+    conn = Http.get_conn(verbose, url)
+    res = conn.get 'status/queue'
+    requests = JSON.parse(res.body)
+
+    ret_val = []
+    requests.each do |req|
+      req_hash = JSON.parse(req)
+      next unless user == req_hash['request']['job']['user']
+
+      ret_val.push(req_hash)
+    end
+
+    ret_val
+  end
+
+  def self.all_job_resources_accounted_for(allocated_resources, hosts)
+    allocated_host_list = allocated_resources.map { |ar| ar['hostname'] }
+    (allocated_host_list - hosts).empty?
+  end
+
+  def self.delete(verbose, url, hosts, token, user)
+    # In ABS terms, this is a "returned" host.
+    conn = Http.get_conn(verbose, url)
+    conn.headers['X-AUTH-TOKEN'] = token if token
+
+    puts "Trying to delete hosts #{hosts}" if verbose
+    requests = get_active_requests(verbose, url, user)
+
+    jobs_to_delete = []
+
+    requests.each do |req_hash|
+      next unless req_hash['state'] == 'allocated' || req_hash['state'] == 'filled'
+
+      req_hash['allocated_resources'].each do |vm_name, _i|
+        if hosts.include? vm_name['hostname']
+          if all_job_resources_accounted_for(job['allocated_resources'], hosts)
```

Comment:
@mikkergimenez Is this `job` variable defined somewhere? When I try using delete it seems to blow up here.
```
➜  vmfloaty git:(add_abs_vm_get) be bin/floaty delete secular-relay.delivery.puppetlabs.net --service abs --trace

bundler: failed to load command: bin/floaty (bin/floaty)
NameError: undefined local variable or method `job' for ABS:Class
  /s/vmfloaty/lib/vmfloaty/abs.rb:90:in `block (2 levels) in delete'
  /s/vmfloaty/lib/vmfloaty/abs.rb:88:in `each'
  /s/vmfloaty/lib/vmfloaty/abs.rb:88:in `block in delete'
  /s/vmfloaty/lib/vmfloaty/abs.rb:85:in `each'
  /s/vmfloaty/lib/vmfloaty/abs.rb:85:in `delete'
  /s/vmfloaty/lib/vmfloaty/service.rb:115:in `delete'
  /s/vmfloaty/lib/vmfloaty.rb:229:in `block (2 levels) in run'
  /Users/highb/.rbenv/versions/2.5.6/lib/ruby/gems/2.5.0/gems/commander-4.4.7/lib/commander/command.rb:182:in `call'
  /Users/highb/.rbenv/versions/2.5.6/lib/ruby/gems/2.5.0/gems/commander-4.4.7/lib/commander/command.rb:153:in `run'
  /Users/highb/.rbenv/versions/2.5.6/lib/ruby/gems/2.5.0/gems/commander-4.4.7/lib/commander/runner.rb:446:in `run_active_command'
  /Users/highb/.rbenv/versions/2.5.6/lib/ruby/gems/2.5.0/gems/commander-4.4.7/lib/commander/runner.rb:68:in `run!'
  /Users/highb/.rbenv/versions/2.5.6/lib/ruby/gems/2.5.0/gems/commander-4.4.7/lib/commander/delegates.rb:15:in `run!'
  /s/vmfloaty/lib/vmfloaty.rb:454:in `run'
  bin/floaty:8:in `<top (required)>'
```